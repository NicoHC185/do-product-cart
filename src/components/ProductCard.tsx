import React from 'react'
import { createContext, CSSProperties, JSX } from 'react'
import { InitialValues, onChangeType, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/product.interface'
import { useProducts } from '../hooks/useProducts'
import styles from '../styles/styles.module.css'

export interface Props {
    product: Product,
    children: (Props: ProductCardHandlers) => JSX.Element,
    className?: string
    style?: CSSProperties
    value?: number
    onChange?: onChangeType
    initialValues?: InitialValues
}

export const ProductContext = createContext({} as ProductContextProps)

const { Provider } = ProductContext

export const ProductCard = ({
    product,
    children,
    className,
    style,
    value,
    onChange,
    initialValues
}: Props) => {

    const { counter, maxCount, increaseBy, isMaxCountReached, reset }
        = useProducts({ product, value, onChange, initialValues })

    return (
        <Provider value={{ counter, maxCount, increaseBy, product }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {children({
                    count: counter,
                    maxCount,
                    product,
                    increaseBy,
                    isMaxCountReached,
                    reset,
                })}
            </div>
        </Provider>
    )
}
