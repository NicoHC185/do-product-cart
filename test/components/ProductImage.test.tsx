import React from 'react'
import { render } from '@testing-library/react';
import ProductCard, { ProductImage } from '../../src/components'
import { product1, product2 } from './data/product';

describe('ProductImage', () => {
    test('Debe de mostrar el componente con la imagen personalizada', () => {
        const customImg = 'https://sm.ign.com/ign_es/news/m/monster-hu/monster-hunter-wilds-mod-already-gives-unlimited-character-a_z9ga.jpg'
        const { container } = render(
            <ProductImage img={customImg} />
        )
        expect(container).toMatchSnapshot()
    })
    test('Debe de mostrar el componente con la imagen del product2', () => {
        const { container }  = render(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )
        expect(container).toMatchSnapshot()

    })
    test('Debe de mostrar el componente con la imagen del product1 (sin imagen)', () => {
        const { container }  = render(
            <ProductCard product={product1}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )
        expect(container).toMatchSnapshot()

    })
})
