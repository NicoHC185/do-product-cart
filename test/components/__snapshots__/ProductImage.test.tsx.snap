// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe de mostrar el componente con la imagen del product1 (sin imagen) 1`] = `
<div>
  <div
    class="productCard undefined"
  >
    <img
      alt="Product img"
      class="productImg undefined"
      src="../assets/no-image.jpg"
    />
  </div>
</div>
`;

exports[`ProductImage Debe de mostrar el componente con la imagen del product2 1`] = `
<div>
  <div
    class="productCard undefined"
  >
    <img
      alt="Product img"
      class="productImg undefined"
      src="./coffe-mug.png"
    />
  </div>
</div>
`;

exports[`ProductImage Debe de mostrar el componente con la imagen personalizada 1`] = `
<div>
  <img
    alt="Product img"
    class="productImg undefined"
    src="https://sm.ign.com/ign_es/news/m/monster-hu/monster-hunter-wilds-mod-already-gives-unlimited-character-a_z9ga.jpg"
  />
</div>
`;
